src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,792 @@
+   mat_iso::GenMatIso
+   order::fmpz
+   AbstractAlgebra.@declare_other
+
+   MatrixGroup{RE,T}(m::Int, F::Ring) where {RE,T} = new{RE,T}(m,F)
+
+end
+
+MatrixGroup(m::Int, F::Ring) = MatrixGroup{elem_type(F), dense_matrix_type(elem_type(F))}(m,F)
+
+# build a MatrixGroup given a list of generators, given as array of either MatrixGroupElem or AbstractAlgebra matrices
+# WARNING: if the elements of V have type MatElem, it does not check whether the determinant is nonzero
+function MatrixGroup{RE,S}(m::Int, F::Ring, V::AbstractVector{T}) where {RE,S} where T<:Union{MatElem,AbstractMatrixGroupElem}
+   G = MatrixGroup(m,F)
+
+   L = Vector{MatrixGroupElem}(undef, length(V))
+   for i in 1:length(V)
+      @assert base_ring(V[i])==F "The elements must have the same base ring"
+      @assert nrows(V[i])==m "The elements must have the same dimension"
+      if T<:MatElem
+         L[i] = MatrixGroupElem(G,V[i])
+      else
+# TODO: this part of code from here
+         if isdefined(V[i],:elm)
+            if isdefined(V[i],:X)
+               L[i] = MatrixGroupElem(G,V[i].elm,V[i].X)
+            else
+               L[i] = MatrixGroupElem(G,V[i].elm)
+            end
+         else
+            L[i] = MatrixGroupElem(G,V[i].X)
+         end
+# to here
+
+# can be replaced by the following line once deepcopy works for GAP objects
+# > L[i] = deepcopy(V[i]); L[i].parent = G;
+      end
+   end
+   G.gens = L
+
+   return G
+end
+
+MatrixGroup(m::Int, F::Ring, V::AbstractVector{T}) where T<:Union{MatElem,AbstractMatrixGroupElem} = MatrixGroup{elem_type(F), dense_matrix_type(elem_type(F))}(m,F,V)
+
+# NOTE: at least one of the fields :elm and :X must always defined, but not necessarily both of them.
+"""
+    MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: GAPGroupElem{MatrixGroup}
+
+Elements of a group of type `MatrixGroup{RE<:RingElem, T<:MatElem{RE}}`
+"""
+mutable struct MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: AbstractMatrixGroupElem
+   parent::MatrixGroup{RE, T}
+   elm::T                         # Oscar matrix
+   X::GapObj                     # GAP matrix. If x isa MatrixGroupElem, then x.X = x.parent.mat_iso(x.elm)
+
+   MatrixGroupElem{RE,T}(G::MatrixGroup{RE,T}, x::T) where {RE, T} = new{RE,T}(G,x)
+
+# build a MatrixGroupElem given both an OSCAR matrix and a corresponding GAP object.
+# WARNING: this does not check whether the element actually lies in the group G
+# nor whether the two given matrices really match.
+   function MatrixGroupElem{RE,T}(G::MatrixGroup{RE,T}, x_gap::GapObj) where {RE, T}
+      z = new{RE,T}()
+      z.parent = G
+      z.X = x_gap
+      return z
+   end
+
+end
+
+function MatrixGroupElem(G::MatrixGroup{RE,T}, x::T, x_gap::GapObj) where {RE,T}
+   z = MatrixGroupElem{RE,T}(G,x)
+   z.X = x_gap
+   return z
+end
+
+function MatrixGroupElem{RE,T}(G::MatrixGroup{RE,T}, x::T, x_gap::GapObj) where {RE,T}
+   z = MatrixGroupElem{RE,T}(G,x)
+   z.X = x_gap
+   return z
+end
```

Comment:
This makes no sense as written. How about this
```suggestion
   # full constructor
   MatrixGroupElem{RE,T}(G::MatrixGroup{RE,T}, x::T, x_gap::GapObj) where {RE, T} = new{RE,T}(G, x, x_gap)

   # constructor which leaves `X` undefined
   MatrixGroupElem{RE,T}(G::MatrixGroup{RE,T}, x::T) where {RE, T} = new{RE,T}(G, x)

   # constructor which leaves `elm` undefined
   function MatrixGroupElem{RE,T}(G::MatrixGroup{RE,T}, x_gap::GapObj) where {RE, T}
      z = new{RE,T}(G)
      z.X = x_gap
      return z
   end

end

MatrixGroupElem(G::MatrixGroup{RE,T}, x::T, x_gap::GapObj) where {RE,T} = MatrixGroupElem{RE,T}(G, x, x_gap)
```